lib/semantics/tools.cc

```diff
@@ -536,6 +537,150 @@ std::unique_ptr<parser::Message> WhyNotModifiable(parser::CharBlock at,
   return {};
 }
 
+static bool IsCoarray(const parser::AllocateObject &allocateObject) {
+  const parser::Name &name{GetLastName(allocateObject)};
+  return name.symbol && IsCoarray(*name.symbol);
+}
+
+static bool IsImageControlStmt(const parser::AllocateStmt &allocateStmt) {
+  const auto &allocationList{
+      std::get<std::list<parser::Allocation>>(allocateStmt.t)};
+  for (const auto &allocation : allocationList) {
+    const auto &allocateObject{std::get<parser::AllocateObject>(allocation.t)};
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool IsImageControlStmt(const parser::DeallocateStmt &deallocateStmt) {
+  const auto &allocateObjectList{
+      std::get<std::list<parser::AllocateObject>>(deallocateStmt.t)};
+  for (const auto &allocateObject : allocateObjectList) {
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool IsImageControlStmt(const parser::CallStmt &callStmt) {
+  const auto &procedureDesignator{
+      std::get<parser::ProcedureDesignator>(callStmt.v.t)};
+  if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+    // TODO: also ensure that the procedure is, in fact, an intrinsic
+    if (name->source == "move_alloc") {
+      const auto &args{
+          std::get<std::list<parser::ActualArgSpec>>(callStmt.v.t)};
+      if (!args.empty()) {
+        const parser::ActualArg &actualArg{
+            std::get<parser::ActualArg>(args.front().t)};
+        if (const auto *argExpr{
+                std::get_if<common::Indirection<parser::Expr>>(&actualArg.u)}) {
+          return HasCoarray(argExpr->value());
+        }
+      }
+    }
+  }
+  return false;
+}
```

Comment:
The four functions above can move into `ImageControlStmtHelper` as that is the only place they are called.
And the three overloads of `IsImageControlStmt` are only called in one place each, so they could be expanded in-line.

Reply:
Good point.  I'll move them into the helper class.